lib/polyfill/pip.js

```diff
@@ -0,0 +1,230 @@
+      if (type === 'enterpictureinpicture') {
+        const proxyEnterPiPEvent =
+          shaka.polyfill.PiP.getProxyEnterPiPEvent_(listener);
+        this.addEventListener('webkitpresentationmodechanged',
+          proxyEnterPiPEvent);
+      } else if (type === 'leavepictureinpicture') {
+        const proxyLeavePiPEvent =
+          shaka.polyfill.PiP.getProxyLeavePiPEvent_(listener);
+        this.addEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldAddEventListener.apply(this, [type, listener, options]);
+      }
+    };
+
+    const oldRemoveEventListener = proto.removeEventListener;
+    /**
+     * @this {HTMLMediaElement}
+     */
+    proto.removeEventListener = function(type, listener, options) {
+      if (type === 'enterpictureinpicture') {
+        const proxyEnterPiPEvent =
+          shaka.polyfill.PiP.getProxyEnterPiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyEnterPiPEvent);
+      } else if (type === 'leavepictureinpicture') {
+        const proxyLeavePiPEvent =
+          shaka.polyfill.PiP.getProxyLeavePiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldRemoveEventListener.apply(this, [type, listener, options]);
+      }
+    };
+  }
+};
+
+/**
+ * @this {HTMLMediaElement}
+ * @private
+ */
+shaka.polyfill.PiP.updatePictureInPictureElementInDocument_ = function() {
+  if (this.webkitPresentationMode &&
+    this.webkitPresentationMode !== 'picture-in-picture') {
+    document.pictureInPictureElement = null;
+  }
+};
+
+/**
+ * @this {HTMLMediaElement}
+ * @return {!Promise}
+ * @private
+ */
+shaka.polyfill.PiP.requestPictureInPicture_ = function() {
+  // check if PIP is enabled
+  if (!this.webkitSupportsPresentationMode('picture-in-picture')) {
+    const error = new Error('PIP not allowed by videoElement',
+      'InvalidStateError');
+    return Promise.reject(error);
+  } else {
+    // enter PIP mode
+    this.webkitSetPresentationMode('picture-in-picture');
+    document.pictureInPictureElement = this;
+    return Promise.resolve();
+  }
+};
+
+/**
+ * @this {Document}
+ * @return {!Promise}
+ * @private
+ */
+shaka.polyfill.PiP.exitPictureInPicture_ = function() {
+  if (document.pictureInPictureElement) {
+    // exit PIP mode
+    const video =
+      /** @type {!HTMLMediaElement} */ (document.pictureInPictureElement);
+    video.webkitSetPresentationMode('inline');
+    document.pictureInPictureElement = null;
+    return Promise.resolve();
+  } else {
+    const error = new Error('No picture in picture element found',
+      'InvalidStateError');
+    return Promise.reject(error);
+  }
+};
+
+/**
+ * Get the proxy webkitpresentationmodechanged event.
+ *
+ * @param {EventListener} listener
```

Comment:
I think the compiler error is because addEventListener has a more complex signature that you would think.  It accepts more than just EventListener, so the argument passed here could have other types, as well.

Could you listen to the webkit events all the time and dispatch equivalent standard events when they are seen?  Like we do in the fullscreen polyfill?

Reply:
Is not the same case than fullscreen. In fullscreen polyfill the event is trigger over the document. In this case, the event is trigger over the video element. I'll try to find another solution.

Reply:
If the only remaining issue is the compiler error on the parameter type here, you could just extend it.

Reply:
Try this:

```js
@param {EventListener|function(Event): (boolean|undefined)} listener
```

Reply:
ERROR - actual parameter 1 of shaka.polyfill.PiP.getProxyLeavePiPEvent_ does not match formal parameter
found   : (EventListener|function(Event): (boolean|undefined)|null)
required: (EventListener|function((Event|null)): (boolean|undefined)|null)